### PR TITLE
Add db-driven mailshots

### PR DESCRIPTION
**Pull request by iHiD**
Doing PRs and deploying the website every time we want to send a mailshot is becoming ridiculous, so I'm moving it into the database. 

## TODO
- [x] Create models
- [x] Create admin section
- [x] Update `user_mailshot` sending code
- [x] Add `mailshot` sending code
- [x] Add magic image link support (via @ErikSchierboom's magic)

### CODE REVIEW

**Comment by iHiD** [comment]
(cc @dem4ron too :))

File: app/commands/mailshot/send_to_audience_segment.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
+class Mailshot
+  class SendToAudienceSegment
+    include Mandate
+
+    initialize_with :mailshot, :audience_type, :audience_slug, :limit, :offset
+
+    # TOOD: This is where we can add throttling to check how many
+    # mails have be sent today and reschedule ourselves for the next
```
Comment (ErikSchierboom) [inline app/commands/mailshot/send_to_audience_segment.rb]:
```suggestion
    # mails have been sent today and reschedule ourselves for the next
```

File: app/commands/mailshot/send_to_audience_segment.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+      records.each do |record|
+        user = user_extractor.(record)
+        next unless user
+
+        User::Mailshot::Send.(user, mailshot)
+      end
+    end
+
+    def schedule_next_segment!
+      Mailshot::SendToAudienceSegment.defer(
```
Comment (ErikSchierboom) [inline app/commands/mailshot/send_to_audience_segment.rb:34]: Do we want to explicitly specify the queue this command runs on?

File: app/models/mailshot.rb [hunk 3]
```diff
@@ -4,4 +4,35 @@ class Mailshot < ApplicationRecord
...
+  def audience_for(type, slug)
+    send("audience_for_#{type}", slug)
+  end
+
+  # rubocop:disable Lint/NonLocalExitFromIterator
+
+  # This is pretty terribly slow and should only be used rarely.
+  def audience_for_admins(_)
+    [
+      User.where("JSON_CONTAINS(roles, ?, '$')", %("admin")),
```
Comment (ErikSchierboom) [inline app/models/mailshot.rb:23]: I spent some time trying to figure out a way to make this fast, but couldn't find it. I was slightly amazed by that

**Review by ErikSchierboom — APPROVED** [review]

File: app/commands/mailshot/send_to_audience_segment.rb (hunk 2 above)
Comment (iHiD) [reply app/commands/mailshot/send_to_audience_segment.rb:34]: Fuck yes. Good man!

File: app/models/mailshot.rb (hunk 3 above)
Comment (iHiD) [reply app/models/mailshot.rb:23]: I didn't acutally check, but I expect it to be terrible. The newer version of MySQL is apparently better, but the one in production will be especially terrible.

Comment (iHiD) [reply app/models/mailshot.rb:23]: All the JSON stuff is bad in this version of MySQL basically.

**Pull request merged**